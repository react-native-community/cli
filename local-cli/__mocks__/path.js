/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

'use strict';

const mockPath = {};

function reset(platform) {
  Object.assign(mockPath, jest.requireActual('path')[platform]);
}

mockPath.mock = {reset};

reset('posix');

module.exports = mockPath;
