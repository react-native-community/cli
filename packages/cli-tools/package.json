{
  "name": "@react-native-community/cli-tools",
  "version": "20.0.1",
  "license": "MIT",
  "main": "build/index.js",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@vscode/sudo-prompt": "^9.0.0",
    "appdirsjs": "^1.2.4",
    "execa": "^5.0.0",
    "find-up": "^5.0.0",
    "launch-editor": "^2.9.1",
    "mime": "^2.4.1",
    "ora": "^5.4.1",
    "picocolors": "^1.1.1",
    "prompts": "^2.4.2",
    "semver": "^7.5.2"
  },
  "devDependencies": {
    "@react-native-community/cli-types": "20.0.1",
    "@types/mime": "^2.0.1",
    "@types/node": "^20.0.0",
    "@types/prompts": "^2.4.4",
    "@types/shell-quote": "^1.7.1"
  },
  "files": [
    "build",
    "!*.d.ts",
    "!*.map"
  ],
  "homepage": "https://github.com/react-native-community/cli/tree/main/packages/cli-tools",
  "repository": {
    "type": "git",
    "url": "https://github.com/react-native-community/cli.git",
    "directory": "packages/tools"
  },
  "gitHead": "3614a01613030c0a328ad2e8bb5d897e5bae1456"
}
