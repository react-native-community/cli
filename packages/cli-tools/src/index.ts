export {default as logger} from './logger';
export {default as groupFilesByType} from './groupFilesByType';
export {default as isPackagerRunning} from './isPackagerRunning';
export {default as getDefaultUserTerminal} from './getDefaultUserTerminal';
export {fetch, fetchToTemp} from './fetch';
export {default as launchDefaultBrowser} from './launchDefaultBrowser';
export {default as launchDebugger} from './launchDebugger';
export {default as launchEditor} from './launchEditor';
export {default as releaseChecker} from './releaseChecker';
export {default as resolveNodeModuleDir} from './resolveNodeModuleDir';
export {default as hookStdout} from './hookStdout';
export {getLoader, NoopLoader, Loader} from './loader';
export {default as findProjectRoot} from './findProjectRoot';

export * from './errors';
