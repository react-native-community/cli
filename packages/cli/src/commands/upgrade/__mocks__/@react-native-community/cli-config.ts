export default function mockedLoadConfig() {
  return {
    root: '/project/root',
    reactNativePath: '',
    commands: [],
    platforms: {
      ios: {projectConfig: () => null, dependencyConfig: () => null},
      android: {projectConfig: () => null, dependencyConfig: () => null},
    },
    project: {
      ios: null,
      android: null,
    },
    dependencies: {},
    assets: [],
  };
}
