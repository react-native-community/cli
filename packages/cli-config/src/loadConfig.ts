import path from 'path';
import {
  UserDependencyConfig,
  ProjectConfig,
  DependencyConfig,
  UserConfig,
  Config,
} from '@react-native-community/cli-types';
import {
  findProjectRoot,
  version,
  resolveNodeModuleDir,
  UnknownProjectError,
} from '@react-native-community/cli-tools';
import findDependencies from './findDependencies';
import resolveReactNativePath from './resolveReactNativePath';
import {
  readConfigFromDisk,
  readDependencyConfigFromDisk,
} from './readConfigFromDisk';
import assign from './assign';
import merge from './merge';

function getDependencyConfig(
  root: string,
  dependencyName: string,
  finalConfig: Config,
  config: UserDependencyConfig,
  userConfig: UserConfig,
  isPlatform: boolean,
): DependencyConfig {
  return merge(
    {
      root,
      name: dependencyName,
      platforms: Object.keys(finalConfig.platforms).reduce(
        (dependency, platform) => {
          const platformConfig = finalConfig.platforms[platform];
          dependency[platform] =
            // Linking platforms is not supported
            isPlatform || !platformConfig
              ? null
              : platformConfig.dependencyConfig(
                  root,
                  config.dependency.platforms[platform],
                );
          return dependency;
        },
        {} as Config['platforms'],
      ),
    },
    userConfig.dependencies[dependencyName] || {},
  ) as DependencyConfig;
}

/**
 * Loads CLI configuration
 */
function loadConfig(projectRoot: string = findProjectRoot()): Config {
  let lazyProject: ProjectConfig;
  const userConfig = readConfigFromDisk(projectRoot);

  const initialConfig: Config = {
    root: projectRoot,
    get reactNativePath() {
      return userConfig.reactNativePath
        ? path.resolve(projectRoot, userConfig.reactNativePath)
        : resolveReactNativePath(projectRoot);
    },
    reactNativeVersion: 'unknown',
    dependencies: userConfig.dependencies,
    commands: userConfig.commands,
    healthChecks: [],
    platforms: userConfig.platforms,
    get project() {
      if (lazyProject) {
        return lazyProject;
      }

      lazyProject = {};
      for (const platform in finalConfig.platforms) {
        const platformConfig = finalConfig.platforms[platform];
        if (platformConfig) {
          lazyProject[platform] = platformConfig.projectConfig(
            projectRoot,
            userConfig.project[platform] || {},
          );
        }
      }

      return lazyProject;
    },
  };

  // Try our best to figure out what version of React Native we're running. This is
  // currently being used to get our deeplinks working, so it's only worried with
  // the major and minor version.
  try {
    let semver = version.current(initialConfig.reactNativePath);
    if (semver) {
      // Retain only these version, since they correspond with our documentation.
      initialConfig.reactNativeVersion = `${semver.major}.${semver.minor}`;
    }
  } catch (e) {
    // If we don't seem to be in a well formed project, give up quietly.
    if (!(e instanceof UnknownProjectError)) {
      throw e;
    }
  }

  const finalConfig = Array.from(
    new Set([
      ...Object.keys(userConfig.dependencies),
      ...findDependencies(projectRoot),
    ]),
  ).reduce((acc: Config, dependencyName) => {
    const localDependencyRoot =
      userConfig.dependencies[dependencyName] &&
      userConfig.dependencies[dependencyName].root;

    try {
      let root =
        localDependencyRoot ||
        resolveNodeModuleDir(projectRoot, dependencyName);
      let config = readDependencyConfigFromDisk(root, dependencyName);

      const isPlatform = Object.keys(config.platforms).length > 0;

      return assign({}, acc, {
        dependencies: assign({}, acc.dependencies, {
          get [dependencyName](): DependencyConfig {
            return getDependencyConfig(
              root,
              dependencyName,
              finalConfig,
              config,
              userConfig,
              isPlatform,
            );
          },
        }),
        commands: [...acc.commands, ...config.commands],
        platforms: {
          ...acc.platforms,
          ...config.platforms,
        },
        healthChecks: [...acc.healthChecks, ...config.healthChecks],
      }) as Config;
    } catch {
      return acc;
    }
  }, initialConfig);

  return finalConfig;
}

export default loadConfig;
