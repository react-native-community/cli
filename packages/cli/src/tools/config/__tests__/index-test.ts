import path from 'path';
import slash from 'slash';
import loadConfig from '..';
import {logger} from '@react-native-community/cli-tools';
import {
  cleanup,
  writeFiles,
  getTempDirectory,
} from '../../../../../../jest/helpers';

jest.mock('../resolveNodeModuleDir');

let DIR = getTempDirectory('config_test');

const iosPath = slash(
  require.resolve('@react-native-community/cli-platform-ios'),
);
const androidPath = slash(
  require.resolve('@react-native-community/cli-platform-android'),
);

const REACT_NATIVE_MOCK = {
  'node_modules/react-native/package.json': '{}',
  'node_modules/react-native/react-native.config.js': `
    const ios = require("${iosPath}");
    const android = require("${androidPath}");
    module.exports = {
      platforms: {
        ios: {
          linkConfig: ios.linkConfig,
          projectConfig: ios.projectConfig,
          dependencyConfig: ios.dependencyConfig,
        },
        android: {
          linkConfig: android.linkConfig,
          projectConfig: android.projectConfig,
          dependencyConfig: android.dependencyConfig,
        },
      },
    };
  `,
};

// Removes string from all key/values within an object
const removeString = (config, str) =>
  JSON.parse(
    JSON.stringify(config, (_key, value) =>
      typeof value === 'string'
        ? slash(value.replace(str, '<<REPLACED>>'))
        : value,
    ),
  );

beforeEach(async () => {
  await cleanup(DIR);
  jest.resetModules();
  jest.clearAllMocks();
});

afterEach(async () => await cleanup(DIR));

test('should have a valid structure by default', () => {
  DIR = getTempDirectory('config_test_structure');
  writeFiles(DIR, {
    'react-native.config.js': `module.exports = {
      reactNativePath: "."
    }`,
  });
  const config = loadConfig(DIR);
  expect(removeString(config, DIR)).toMatchSnapshot();
});

test('should return dependencies from package.json', () => {
  DIR = getTempDirectory('config_test_deps');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ios/HelloWorld.xcodeproj/project.pbxproj':
      '',
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(removeString(dependencies, DIR)).toMatchSnapshot();
});

test('should read a config of a dependency and use it to load other settings', () => {
  DIR = getTempDirectory('config_test_settings');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ReactNativeTest.podspec': '',
    'node_modules/react-native-test/react-native.config.js': `module.exports = {
      dependency: {
        platforms: {
          ios: {
            project: "./customLocation/customProject.xcodeproj"
          }
        }
      }
    }`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(
    removeString(dependencies['react-native-test'], DIR),
  ).toMatchSnapshot();
});

test('should merge project configuration with default values', () => {
  DIR = getTempDirectory('config_test_merge');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/react-native.config.js': `module.exports = {
      dependency: {
        assets: ["foo", "baz"]
      }
    }`,
    'node_modules/react-native-test/ios/HelloWorld.xcodeproj/project.pbxproj':
      '',
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
    'react-native.config.js': `module.exports = {
      reactNativePath: ".",
      dependencies: {
        "react-native-test": {
          platforms: {
            ios: {
              sourceDir: "./abc"
            }
          },
          assets: ["foo"]
        }
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(removeString(dependencies['react-native-test'], DIR)).toMatchSnapshot(
    'snapshoting `react-native-test` config',
  );
});

test('should load commands from "react-native-foo" and "react-native-bar" packages', () => {
  DIR = getTempDirectory('config_test_packages');
  writeFiles(DIR, {
    'node_modules/react-native-foo/package.json': '{}',
    'node_modules/react-native-foo/react-native.config.js': `module.exports = {
      commands: [
        {
          name: 'foo-command',
          func: () => console.log('foo')
        }
      ]
    }`,
    'node_modules/react-native-bar/package.json': '{}',
    'node_modules/react-native-bar/react-native.config.js': `module.exports = {
      commands: [
        {
          name: 'bar-command',
          func: () => console.log('bar')
        }
      ]
    }`,
    'package.json': `{
      "dependencies": {
        "react-native-foo": "0.0.1",
        "react-native-bar": "0.0.1"
      }
    }`,
  });
  const {commands} = loadConfig(DIR);
  expect(commands).toMatchSnapshot();
});

test('should skip packages that have invalid configuration', () => {
  DIR = getTempDirectory('config_test_skip');
  writeFiles(DIR, {
    'node_modules/react-native/package.json': '{}',
    'node_modules/react-native/react-native.config.js': `module.exports = {
      dependency: {
        invalidProperty: 5
      }
    }`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1"
      }
    }`,
  });
  const spy = jest.spyOn(logger, 'warn');
  const {dependencies} = loadConfig(DIR);
  expect(dependencies).toMatchSnapshot('dependencies config');
  expect(spy.mock.calls[0][0]).toMatchSnapshot('logged warning');
});

test('does not use restricted "react-native" key to resolve config from package.json', () => {
  DIR = getTempDirectory('config_test_restricted');
  writeFiles(DIR, {
    'node_modules/react-native-netinfo/package.json': `{
      "react-native": "src/index.js"
    }`,
    'package.json': `{
      "dependencies": {
        "react-native-netinfo": "0.0.1"
      }
    }`,
  });
  const spy = jest.spyOn(logger, 'warn');
  const {dependencies} = loadConfig(DIR);
  expect(dependencies).toHaveProperty('react-native-netinfo');
  expect(spy).not.toHaveBeenCalled();
});

test('supports dependencies from user configuration with custom root and properties', () => {
  DIR = getTempDirectory('config_test_custom_root');
  const escapePathSeparator = (value: string) =>
    path.sep === '\\' ? value.replace(/(\/|\\)/g, '\\\\') : value;

  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'native-libs/local-lib/ios/LocalRNLibrary.xcodeproj/project.pbxproj': '',
    'react-native.config.js': `
const path = require('path');
const root = path.resolve('${escapePathSeparator(
      DIR,
    )}', 'native-libs', 'local-lib');

module.exports = {
  dependencies: {
    'local-lib': {
      root,
      platforms: {
        ios: {
          podspecPath: "custom-path"
        }
      }
    },
  }
}`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1"
      }
    }`,
  });

  const {dependencies} = loadConfig(DIR);
  expect(removeString(dependencies['local-lib'], DIR)).toMatchInlineSnapshot(`
    Object {
      "assets": Array [],
      "hooks": Object {},
      "name": "local-lib",
      "params": Array [],
      "platforms": Object {
        "android": null,
        "ios": Object {
          "configurations": Array [],
          "folder": "<<REPLACED>>/native-libs/local-lib",
          "libraryFolder": "Libraries",
          "pbxprojPath": "<<REPLACED>>/native-libs/local-lib/ios/LocalRNLibrary.xcodeproj/project.pbxproj",
          "plist": Array [],
          "podfile": null,
          "podspecPath": "custom-path",
          "projectName": "LocalRNLibrary.xcodeproj",
          "projectPath": "<<REPLACED>>/native-libs/local-lib/ios/LocalRNLibrary.xcodeproj",
          "scriptPhases": Array [],
          "sharedLibraries": Array [],
          "sourceDir": "<<REPLACED>>/native-libs/local-lib/ios",
        },
      },
      "root": "<<REPLACED>>/native-libs/local-lib",
    }
  `);
});

test('should apply build types from dependency config', () => {
  DIR = getTempDirectory('config_test_apply_dependency_config');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ios/HelloWorld.xcodeproj/project.pbxproj':
      '',
    'node_modules/react-native-test/react-native.config.js': `module.exports = {
      dependency: {
        platforms: {
          ios: {
            configurations: ["debug"]
          }
        }
      }
    }`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(
    removeString(dependencies['react-native-test'], DIR),
  ).toMatchSnapshot();
});

test('supports dependencies from user configuration with custom build type', () => {
  DIR = getTempDirectory('config_test_apply_custom_build_config');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'react-native.config.js': `module.exports = {
  dependencies: {
    'react-native-test': {
      platforms: {
        ios: {
          configurations: ["custom_build_type"]
        }
      }
    },
  }
}`,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ios/HelloWorld.xcodeproj/project.pbxproj':
      '',
    'node_modules/react-native-test/react-native.config.js':
      'module.exports = {}',
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });

  const {dependencies} = loadConfig(DIR);
  expect(
    removeString(dependencies['react-native-test'], DIR),
  ).toMatchSnapshot();
});
