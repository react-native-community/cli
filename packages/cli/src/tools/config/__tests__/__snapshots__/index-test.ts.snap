// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should automatically put "react-native" into haste config 1`] = `
Object {
  "platforms": Array [
    "ios",
    "android",
  ],
  "providesModuleNodeModules": Array [
    "react-native",
  ],
}
`;

exports[`should have a valid structure by default 1`] = `
Object {
  "assets": Array [],
  "commands": Array [],
  "dependencies": Object {},
  "haste": Object {
    "platforms": Array [],
    "providesModuleNodeModules": Array [],
  },
  "platforms": Object {},
  "project": Object {},
  "reactNativePath": "<<REPLACED>>",
  "root": "<<REPLACED>>",
}
`;

exports[`should load an out-of-tree "windows" platform that ships with a dependency 1`] = `
Object {
  "haste": Object {
    "platforms": Array [
      "windows",
    ],
    "providesModuleNodeModules": Array [
      "react-native-windows",
    ],
  },
  "platforms": Object {
    "windows": Object {},
  },
}
`;

exports[`should load commands from "react-native-foo" and "react-native-bar" packages 1`] = `
Array [
  Object {
    "func": [Function],
    "name": "foo-command",
  },
  Object {
    "func": [Function],
    "name": "bar-command",
  },
]
`;

exports[`should merge project configuration with default values: snapshoting \`react-native-test\` config 1`] = `
Object {
  "name": "react-native-test",
  "platforms": Object {
    "android": null,
    "ios": Object {
      "podspecPath": null,
      "scriptPhases": Array [],
      "sourceDir": "./abc",
    },
  },
  "root": "<<REPLACED>>/node_modules/react-native-test",
}
`;

exports[`should not add default React Native config when one present 1`] = `
Array [
  Object {
    "func": [Function],
    "name": "test",
  },
]
`;

exports[`should read a config of a dependency and use it to load other settings 1`] = `
Object {
  "name": "react-native-test",
  "platforms": Object {
    "android": null,
    "ios": Object {
      "podspecPath": "<<REPLACED>>/node_modules/react-native-test/ReactNativeTest.podspec",
      "scriptPhases": Array [],
      "sourceDir": "<<REPLACED>>/node_modules/react-native-test/customLocation",
    },
  },
  "root": "<<REPLACED>>/node_modules/react-native-test",
}
`;

exports[`should return dependencies from package.json 1`] = `
Object {
  "react-native": Object {
    "name": "react-native",
    "platforms": Object {
      "android": null,
      "ios": null,
    },
    "root": "<<REPLACED>>/node_modules/react-native",
  },
  "react-native-test": Object {
    "name": "react-native-test",
    "platforms": Object {
      "android": null,
      "ios": Object {
        "podspecPath": null,
        "scriptPhases": Array [],
        "sourceDir": "<<REPLACED>>/node_modules/react-native-test/ios",
      },
    },
    "root": "<<REPLACED>>/node_modules/react-native-test",
  },
}
`;
